oak-blob-cloud/src/main/java/org/apache/jackrabbit/oak/blob/cloud/s3/S3RequestDecorator.java

```diff
@@ -58,20 +88,63 @@ public PutObjectRequest decorate(PutObjectRequest request) {
      * Set encryption in {@link CopyObjectRequest}
      */
     public CopyObjectRequest decorate(CopyObjectRequest request) {
+        ObjectMetadata metadata;
         switch (getDataEncryption()) {
             case SSE_S3:
-                ObjectMetadata metadata = request.getNewObjectMetadata() == null
+                metadata = request.getNewObjectMetadata() == null
                                 ? new ObjectMetadata()
                                 : request.getNewObjectMetadata();
                 metadata.setSSEAlgorithm(ObjectMetadata.AES_256_SERVER_SIDE_ENCRYPTION);
                 request.setNewObjectMetadata(metadata);
                 break;
+            case SSE_KMS:
+                metadata = request.getNewObjectMetadata() == null
+                                ? new ObjectMetadata()
+                                : request.getNewObjectMetadata();
+                metadata.setSSEAlgorithm(SSEAlgorithm.KMS.getAlgorithm());
+                request.setNewObjectMetadata(metadata);
+                String keyId = getSSEParams().getAwsKmsKeyId();
+                request.withSSEAwsKeyManagementParams(keyId == null
+                        ? new SSEAwsKeyManagementParams()
+                        : new SSEAwsKeyManagementParams(keyId));
+                break;
+            case NONE:
+                break;
+        }
+        return request;
+    }
+
+    public InitiateMultipartUploadRequest decorate(InitiateMultipartUploadRequest request) {
+        ObjectMetadata metadata;
+        switch (getDataEncryption()) {
+            case SSE_S3:
+                metadata = request.getObjectMetadata() == null
+                                ? new ObjectMetadata()
+                                : request.getObjectMetadata();
+                metadata.setSSEAlgorithm(ObjectMetadata.AES_256_SERVER_SIDE_ENCRYPTION);
+                request.setObjectMetadata(metadata);
+                break;
+            case SSE_KMS:
+                metadata = request.getObjectMetadata() == null
+                                ? new ObjectMetadata()
+                                : request.getObjectMetadata();
+                metadata.setSSEAlgorithm(SSEAlgorithm.KMS.getAlgorithm());
+                request.setObjectMetadata(metadata);
+                String keyId = getSSEParams().getAwsKmsKeyId();
+                request.withSSEAwsKeyManagementParams(keyId == null
+                        ? new SSEAwsKeyManagementParams()
+                        : new SSEAwsKeyManagementParams(keyId));
```

Comment:
I'd prefer to see a separate method for setting up the appropriate type of metadata in the request, so it is just a simple call from each case, e.g. "request = setRequestEncryptionSSE_KMS(request)" or something along those lines.  This way we isolate this logic in a single place.